src/standalone/Soil/rre.jl

```diff
@@ -420,57 +415,71 @@ to using the modified Picard scheme of Celia et al. (1990).
         gradc2f_op = Operators.GradientC2F(
             top = Operators.SetGradient(Geometry.WVector.(FT(0))),
             bottom = Operators.SetGradient(Geometry.WVector.(FT(0))),
         )
-        gradc2f_stencil = Operators.Operator2Stencil(gradc2f_op)
+        gradc2f_matrix = MatrixFields.operator_matrix(gradc2f_op)
+        # Create interpolation operator
         interpc2f_op = Operators.InterpolateC2F(
             bottom = Operators.Extrapolate(),
             top = Operators.Extrapolate(),
         )
-        compose = Operators.ComposeStencils()
 
-        @. ∂ϑₜ∂ϑ = compose(
-            divf2c_stencil(Geometry.Covariant3Vector(W.ones_face_space)),
-            (
-                interpc2f_op(p.soil.K) * ClimaLand.to_scalar_coefs(
-                    gradc2f_stencil(
+        # The derivative of the residual with respect to the prognostic variable
+        ∂ϑres∂ϑ =
+            matrix[MatrixFields.@name(soil.ϑ_l), MatrixFields.@name(soil.ϑ_l)]
+        # If the top BC is a `MoistureStateBC`, add the term from the top BC
+        #  flux before applying divergence
+        if haskey(p, :dfluxBCdY)
+            dfluxBCdY = p.dfluxBCdY
+            # TODO not sure if this usage is correct
+            topBC_op = Operators.SetBoundaryOperator(
+                top = Operators.SetValue(Geometry.WVector(dfluxBCdY)),
+                bottom = Operators.SetValue(Geometry.WVector(zero(FT))),
+            )
+            # Add term from top boundary condition before applying divergence
+            @. ∂ϑres∂ϑ =
+                divf2c_matrix() ⋅ (
+                    MatrixFields.DiagonalMatrixRow(interpc2f_op(p.soil.K)) ⋅
+                    gradc2f_matrix() ⋅ MatrixFields.DiagonalMatrixRow(
                         ClimaLand.Soil.dψdϑ(
                             hydrology_cm,
                             Y.soil.ϑ_l,
                             ν,
                             θ_r,
                             S_s,
                         ),
-                    ),
+                    ) + MatrixFields.BidiagonalMatrixRow(topBC_op)
                 )
-            ),
-        )
+        else
+            @. ∂ϑres∂ϑ =
+                divf2c_matrix() ⋅
+                MatrixFields.DiagonalMatrixRow(interpc2f_op(p.soil.K)) ⋅
+                gradc2f_matrix() ⋅ MatrixFields.DiagonalMatrixRow(
+                    ClimaLand.Soil.dψdϑ(hydrology_cm, Y.soil.ϑ_l, ν, θ_r, S_s),
+                )
+
+        end
+
+
         # Hardcoded for single column: FIX!
         # The derivative of the tendency may eventually live in boundary vars and be updated there. but TBD
         z = Fields.coordinate_field(axes(Y.soil.ϑ_l)).z
         Δz_top, Δz_bottom = get_Δz(z)
-        ∂T_bc∂YN = ClimaLand.∂tendencyBC∂Y(
-            model,
-            model.boundary_conditions.top,
-            ClimaLand.TopBoundary(),
-            Δz_top,
-            Y,
-            p,
-            t,
-        )
+
+
+        ∂T_bc∂YN = p.dfluxBCdY
```

Comment:
I believe we can get rid of this + code that follows as we added the BC contribution above, but we can discuss!